### ISSUE TITLE
Scratchpad time-memory tradeoff estimation

### ISSUE THREAD

**Issue by SChernykh** [tevador/RandomX#65]
I've tried to come up with a strategy when least overwritten part of scratchpad (everything above L2) is not stored entirely. Instead it's split into 16-byte chunks. Chunks 0, 2, 4, 6, ... are stored directly and chunks 1, 3, 5, 7, ... are stored only when they're changed. This requires one AES round to read odd chunks and also one level of indirection when reading odd chunks that were changed (indexing). I've run a quick test to see how many odd 16-byte chunks are changed during program executions. Here are my numbers:

low bytes (0 - 262143): 262144 bytes stored directly
high bytes: (2097152 - 262144) / 2 = 917504 bytes stored directly, 917504 bytes are stored as 57344 16-byte chunks and only when changed

99% of hashes have fewer than ~~43400~~ 48100 (of 57344) changed 16-byte chunks (~~694400~~ 769600 bytes) after 8 program executions.
917504 / 16 = 57344, so 16 bits (2 bytes per chunk) is enough for indexing
57344*2 = 114688 bytes for indices

Overall memory usage: 262144 (L2) + 917504 (L3 stored directly) + ~~694400~~ 769600 (L3 stored when changed) + 114688 (L3 indices) = ~~1988736~~ 2063936 bytes = ~~94.83%~~ 98.42% of the original scratchpad

~~Conclusion: it's possible to shave off 5% from each scratchpad at the cost of 1% invalid hashes and also (maybe) some slowdown when reading/writing scratchpad memory above first 262144 bytes.~~

I think it's possible to handle indexing and AES without a slowdown in ASIC, so net performance increase would be 4% because more scratchpads can fit on chip.

Is this speedup something we can simply ignore because it makes ASIC memory logic much more complex, or should we address it with more writes to L3 part of the scratchpad?

**Edit: Forget it, it won't work because even chunks (0,2,4,6 ,..). can also change and it won't be possible to use 1 AES round to read corresponding odd chunks (1,3,5,7, ...), so this will require even more memory because each changed even chunk means we need to store corresponding odd chunk too.**

**Comment by tevador** [comment]
> it won't work because even chunks (0,2,4,6 ,..). can also change and it won't be possible to use 1 AES round to read corresponding odd chunks

There could be a 1-bit 'modified' flag attached to each even chunk and you'd have to potentially make more than 1 AES round by going backwards until you find an unchanged one. This would make reading potentially very slow. Or just abandon the even-odd scheme and have some static checkpoints that always save the original scratchpad value each N rounds.

Secondly, [fully associative](https://en.wikipedia.org/wiki/Cache_placement_policies#Fully_Associative_Cache) cache is impossible to use for more than a few dozen items. So it would need to be a set-associative cache, in which case you'll sometimes have to abort the hash if you run out of space in just one set even if other sets still have unused slots. So the amount of failed hashes would be higher than 1%.

Additionally, reading from cache is slower and requires more energy than reading from scratchpad RAM.

Overall I don't think it's a viable strategy. We have already made this more difficult by making ISTORE write into L3.

**Comment by SChernykh** [comment]
> Secondly, fully associative cache is impossible to use for more than a few dozen items. So it would need to be a set-associative cache

No, this is not a cache I'm describing. Cache accesses items by their addresses and it's a waste of resources in this case. The strategy I describe is still direct addressing, it just does
```
data = scratchpad[index[i]];
```
instead of
```
data = scratchpad[i];
```
Indices are 16-bit wide here.

**Comment by tevador** [comment]
OK, I misunderstood.

Still you need 2 SRAM accesses instead of 1, handle the situation when a chunk is not indexed yet and perform 1 or more AES rounds to recover unmodified chunks. It seems like this should be enough to negate the 2% size reduction if we focus on performance per watt.

**Comment by SChernykh** [comment]
Yes, I agree. The main purpose of this github issue is to find some viable memory reduction strategy, but it seems that we have enough random writes to L3 already to make it impossible. And we also have numbers now to prove it.

**Comment by cjdelisle** [comment]
Assuming for a moment that the scratchpad starts off as zero, we can evaluate just the write side. If my understanding is correct, we have:

* `2**4` ISTORE ops per execution
* `2**14` total executions
* `2**-3` chance of hitting L3

Therefore there should be an average of `2**15` writes to L3, L3 being 229376 uint64 slots, the question becomes whether it can be cheaper to use a cache of `2**15` uint64 slots rather than SRAM of 229376 slots. In terms of silicon real estate, the cost should be comparable to a TLB of `2**15` entries since a TLB is a simple uint64 -> uint64 map, in fact a TLB is probably the best IP core to use for building such a cache.

**What is the die realestate per entry of a TLB compared to SRAM ?**

Lets play with the assumption that `2**15` entry TLB costs us only the same die space as `1**16` uint64 of SRAM. This is probably low but it's a toy assumption. In this case we can better then half our scratchpad by adopting this technique.

## But we don't start with a null scratchpad

As @SChernykh points out in the beginning, we can half our scratchpad by accepting to re-run one AES cycle in case of a "cache miss".

The instructions which are able to pull from L3 are:

* IXOR_M (5/256)
* ISMULH_M (1/256)
* IMULH_M (1/256)
* IMUL_M (4/256)
* ISUB_M (7/256)
* IADD_M (7/256)

But only in the case that the source and dest register are the same, there being 8 registers that should be a 1/8 chance out of a 25/256 chance, or a 25/2048 chance of a load from L3. This number will be slightly lower even because L3 encompasses L2 and L1, but we won't worry about that for now.

Out of `256 * 2**14` instruction executions, we have a 25/2048 chance of an L3 read hit, so we should expect 51200 hits to L3. So now let us assume that no writes to L3 ever happen. To throw away half of L3 we would incur an additional `51200/2` 16 byte AES operations, or about 1/5 additional AES cost. This is somewhat worrying, but we need to consider it in tandem with the stores to L3.

## The chance of reading what you wrote
In principle, "write only" memory is asynchronous and thus limited only by bandwidth. Whatever memory location we write but never read back can be written to slow DRAM to be loaded back later when computing the fingerprint of the entire scratchpad.

Suppose rather than a TLB-like structure, we simply stored a bitset of 229376 bits where 1 indicates modified and 0 indicates unmodified. When doing a scratchpad read, we check the bitset for the relevant uint64 and if it has been written, we accept to go to slow memory for the value but if it is unwritten, we compute the value with AES.

Since I'm out of my depth in terms of calculating probabilities here, I wrote a little python script to simulate hits:

```python
import random

def compute():
    # scratchpad has 2**18 64 bit slots
    scratchpad = {}
    hitCount = 0
    for i in range(0, 2**14 * 2**8):
        x = random.randint(0, 0xffffffff)
        if (x & 0xff) < 2**4:
            # istore instruction
            x >>= 8
            scratchpad[x & ((2**18) - 1)] = True
            continue
        x -= 2**4
        if (x & 0xff) >= 25:
            continue
        # *_M instruction
        x >>= 8
        if (x & 7) > 0:
            continue
        # *_M instruction with src and dest the same (loads from L3)
        x >>= 3
        slot = x & ((2**18) - 1)
        if slot in scratchpad:
            hitCount += 1
    return hitCount

print(str(compute()))
```

What I find is that results hover around 19000, so we can imagine that out of 16k executions, we have less than 2 memory lookups needed per execution and the rest can be computed. These memory lookups cannot be prefetched so they will cause execution to stall, but hyperthreading will probably smooth over this latency.

## In conclusion
On processors with less cache per core but with hardware AES, it is probably more efficient to halve L3 and replace it with a "touched" bitset and then go to slower memory for the times when a memory location is modified and then read back. This issue would be mitigated by reducing L3 such that every address in L3 has a 50% chance of being written.

**Comment by tevador** [comment]
> This issue would be mitigated by reducing L3 such that every address in L3 has a 50% chance of being written

The least written parts of L3 already have a ~66% chance of being overwritten. See: https://github.com/tevador/RandomX/blob/master/doc/design.md#282-scratchpad-writes

**Comment by cjdelisle** [comment]
Ahh yes, sorry about that, I missed the write out of the registers post-computation, then I reverse my conclusion and I don't think any time/memory attack along these lines makes any sense.
